packages/server/src/internals/middlewares.ts

```diff
@@ -0,0 +1,31 @@
+import { ProcedureType } from '../router';
+import { TRPCError } from '../TRPCError';
+
+export const middlewareMarker = Symbol('middlewareMarker');
+interface MiddlewareResultBase {
+  /**
+   * All middlewares should pass through their `next()`'s output.
+   * Requiring this marker makes sure that can't be forgotten at compile-time.
+   */
+  readonly marker: typeof middlewareMarker;
+}
+
+interface MiddlewareOKResult extends MiddlewareResultBase {
```

Comment:
Re the discriminated union idea:

I'm totally open to it, but as a user it might confuse me a little. If my middleware detects an error state, should I `throw ...` or should I return `{ ok: false, error: ... }`? And whatever the "correct" answer is, we can be sure that someone on my team will do it the other way anyway. I _personally_ prefer functional-style and pure code without `throw`s. But `throw` is a language feature that isn't going anywhere, so adding an error type to the middleware API increases the cognitive overhead. We can never get rid of the possibility that a middleware function might throw.

If it's for the sake of the "recoverable" error scenario, IMHO we could keep it simple and save for a future change if necessary (a nice thing about this implementation is it's backwards compatible). Especially since users are already able to "recover" via `catch`.

Reply:
I reverted the discriminated union, but I've not convinced myself yet, will let it simmer a bit; I'm not in a rush to get this in and I want to stop releasing so many majors.

Reply:
What if we set it up so that a breaking change wasn't required, maybe by refactoring `middlewareMarker` to `middlewareSuccessMarker`, leaving us open to adding a `middlewareFailureMarker` later, so the types now would be:

```ts
export const middlewareSuccess = Symbol('middlewareSuccess')

export interface MiddlewareResult {
  readonly marker: typeof middlewareSuccess
  data: unknown
}
```

And if you/the community decide to add in failure return values too, it could become:

```ts
export const middlewareSuccess = Symbol('middlewareSuccess')
export const middlewareFailure = Symbol('middlewareFailure')

interface MiddlewareSuccessResult {
  readonly marker: typeof middlewareSuccess
  data: unknown
}

interface MiddlewareFailureResult {
  readonly marker: typeof middlewareFailure
  error: TRPCError
}

export type MiddlewareResult = MiddlewareSuccessResult | MiddlewareFailureResult
```

Which is a non-breaking change, because all the middlewares out there in the wild would be returning `{ marker: middlewareSuccess, data: ... }` which is still valid. So wouldn't require a major bump.

Plus, I think it's marginally better having both a `marker` and an `ok` when only one is needed.